src/main/java/seedu/address/logic/commands/BuyCommand.java

```diff
@@ -50,6 +61,30 @@ public BuyCommand(Good boughtGood) {
 
     @Override
     public CommandResult execute(Model model) throws CommandException {
+        // supplier could be absent in the model
+        Name supplierNameToFind = boughtGood.getSupplierName();
+        int numMatchingSuppliers = (int) model.getFilteredSupplierList().stream()
+                .filter(supplier -> supplier.getName().equals(supplierNameToFind)).count();
+        if (numMatchingSuppliers <= 0) {
+            throw new CommandException(MESSAGE_BUYING_FROM_NONEXISTENT_SUPPLIER);
+        }
+
+        // good could be absent in the model
+        // currently takes the first supplier with matching Name if there are >1 such suppliers
+        Supplier foundSupplier = model.getFilteredSupplierList().stream()
+                .filter(supplier -> supplier.getName().equals(supplierNameToFind))
+                .findFirst()
+                .get();
+
+        int numMatchingGoods = (int) foundSupplier.getOffers().stream()
+                .filter(offer -> offer.getGoodName()
+                .equals(boughtGood.getGoodName()))
+                .count();
+
+        if (numMatchingGoods <= 0) {
+            throw new CommandException(MESSAGE_SUPPLIER_DOES_NOT_SELL_GOOD);
+        }
+
         if (model.hasGood(boughtGood)) {
             increaseQuantity(model, boughtGood);
         } else {
```

Comment:
Maybe need to add a new BuyTransaction and add it into the TransactionHistory?

Reply:
Having BuyCommand add a BuyTransaction to the transaction history was is the next enhancement after this one is approved.